### DISCUSSION TITLE
How to make columns optional?

### DISCUSSION THREAD

**Discussion by Tar8117** [piccolo-orm/piccolo#606] [Q&A]
Hi !
I have a table.

```
from piccolo.columns import Boolean, Integer, Varchar, Email
from piccolo.table import Table


class Student(Table):
    last_name = Varchar(null=True, length=100)
    first_name = Varchar(null=True, length=100)
    phone_number = Varchar(null=True, length=100, unique=True)
    personal_email = Email(null=True, unique=True)
    cohort_number = Integer(null=True)
    tg_id = Integer(null=True, unique=True)
    tg_login = Varchar(null=True, unique=True)
    begin_work_with_bot = Boolean()
    communication_info = Boolean()
    communication_teamwork = Boolean()
    communication_neformal = Boolean()
    communication_feedback = Boolean()
```
So, I want my columns to be optional, when creating new fields in my table. For example, I want to create empty fields. 
BTW, when I used 0.84.0 version, and didn't use new Email column type, I could create empty fields in my table, without additional parameters (as null = True and etc.), but now I'm trying to make columns optional by using `null = True`, but it doesn't work. Is there any way to make columns optional?

**Comment by dantownsend** [comment]
When you change a column to have `null=True`, make sure you create a migration and apply it.

One important thing to be aware of is columns like `Varchar` (and derived columns like `Email`) have a default value of `''` (i.e. an  empty string).

So if you want  value of `null` if no value is specified, use this:

```python
class MyTable(Table):
    my_column = Varchar(null=True, default=None)
```

Does that help?

**Reply by Tar8117** [reply]
Hmm, no it doesn't help. I noticed smth. You know, when I'm adding `personal_email = Email()`, it effects on Integer type lines. _**cohort_number**_ and _**tg_id**_  have a 0 value by default, and in admin panel there is a 0 by default too. But when I add `personal_email = Email()`, _**cohort_number**_ and _**tg_id**_ become empty (don't contain 0 value by default and are empty) and when trying to create a new field it returns TypeError. So it thinks that I am giving a string instead of integer. I don't know why this is happening.  I tried to give default value visibly `tg_id = Integer(default=0)`, but it doesn't help. Maybe auto validation in Email column type somehow effects on Integer column type?

**Reply by dantownsend** [reply]
OK - maybe there's a bug with Piccolo Admin and `Email` columns. I'll have a look.

**Reply by dantownsend** [reply]
It seems to work OK for me.

You might need to install Pydantic's email support -  `pip install pydantic[email]`

**Reply by dantownsend** [reply]
Are you getting any errors in the Javascript console?

**Reply by Tar8117** [reply]
I'm getting error in admin panel. Above the form of creating table data. JSON response with an error. 
I only installed the new version of piccolo. But ok I will now install as you say and try again.
**UPD:** I have already installed pydantic[email] (Requirement already satisfied: pydantic[email])
Ok I will try again later in a clear new test project. Maybe smth wrong with my project. Thank's for help

**Reply by dantownsend** [reply]
No problem. If you still have an issue, paste the error messages here, and I might be able to figure it out.

**Reply by Tar8117** [reply]
Hi! I tried it again and make some screenshots to show you. If you will have some suggestions please reply. 
P.S.: I make migrations and run them after I changed column types in **table.py** every time. And I make migrations clearly. I mean I delete old migration and table to make it clear.
First 3 screenshots show admin panel with using email column type. 
Another screenshots show correct working when I don't use email column type. 
 
![image](https://user-images.githubusercontent.com/80065505/187911693-e26f050b-12ca-479d-a409-e353733e5ad2.jpeg)
![image](https://user-images.githubusercontent.com/80065505/187911853-17a28c74-43c3-44f2-991a-6129d00e6665.jpeg)
![image](https://user-images.githubusercontent.com/80065505/187911917-ae6605af-cc61-48be-84b8-888540b3aad2.jpeg)
![image](https://user-images.githubusercontent.com/80065505/187911928-86f031e0-d39c-420e-b6b3-81e6cb1ba390.jpeg)
![image](https://user-images.githubusercontent.com/80065505/187911948-49a0274b-f8eb-4318-a884-a5cf496c5872.jpeg)

**Reply by dantownsend** [reply]
Thanks for posting the screenshots - it definitely helps. I'll give it a go.

**Reply by dantownsend** [reply]
Can you paste your code for the `Student` table from `tables.py`?

**Reply by Tar8117** [reply]
Yes of course.

**_table.py_** (when using Email column type):

```
from piccolo.columns import Boolean, Integer, Varchar, BigInt, Email
from piccolo.table import Table


class Student(Table):
    last_name = Varchar(length=100)
    first_name = Varchar(length=100)
    phone_number = Varchar(unique=True, length=100)
    personal_email = Email()
    cohort_number = Integer()
    tg_id = BigInt(unique=True)
    tg_login = Varchar(unique=True)
    begin_work_with_bot = Boolean()
    communication_info = Boolean()
    communication_teamwork = Boolean()
    communication_neformal = Boolean()
    communication_feedback = Boolean()
```

**_table.py_** (when do not using Email column type):
```

from piccolo.columns import Boolean, Integer, Varchar, BigInt
from piccolo.table import Table


class Student(Table):
    last_name = Varchar(length=100)
    first_name = Varchar(length=100)
    phone_number = Varchar(unique=True, length=100)
    personal_email = Varchar(unique=True)
    cohort_number = Integer()
    tg_id = BigInt(unique=True)
    tg_login = Varchar(unique=True)
    begin_work_with_bot = Boolean()
    communication_info = Boolean()
    communication_teamwork = Boolean()
    communication_neformal = Boolean()
    communication_feedback = Boolean()

```

**Reply by dantownsend** [reply]
You're right - there is a bug with Piccolo Admin, sorry about that.

I'll see if I can fix it.

**Reply by dantownsend** [reply]
You can work around for it for now by changing it to this:

```python
personal_email = Email(default='test@test.com')
```

**Reply by dantownsend** [reply]
Or just use `Varchar` until we get it properly fixed.

**Reply by Tar8117** [reply]
Thank you very much

**Reply by dantownsend** [reply]
If you install the latest version of Piccolo and Piccolo API, the problem should be fixed.

When running your code I noticed one thing you should be aware of - you have a few unique columns. If you get an 'Internal Server Error' it's because of a unique constraint (i.e. adding a row with a value which already exists). We're working on showing a proper error message for unique constraint errors.

**Reply by dantownsend** [reply]
> When running your code I noticed one thing you should be aware of - you have a few unique columns. If you get an 'Internal Server Error' it's because of a unique constraint (i.e. adding a row with a value which already exists). We're working on showing a proper error message for unique constraint errors.

If you install the latest version of `piccolo_api` this is now fixed.

**Reply by Tar8117** [reply]
Thank you a lot. I will try today
